Nullptr crash in URLPatternUtilities::URLPatternComponent::createComponentMatchResult

pratiksha-99 opened the pull request:
#### 20ebe20f75010b63e78b719250a9b9a5e9333140
<pre>
Nullptr crash in URLPatternUtilities::URLPatternComponent::createComponentMatchResult
<a href="https://bugs.webkit.org/show_bug.cgi?id=291924">https://bugs.webkit.org/show_bug.cgi?id=291924</a>
<a href="https://rdar.apple.com/149686034">rdar://149686034</a>

Reviewed by Youenn Fablet.

Added a check for null global object context in case it is broken or detached.

* LayoutTests/fast/url/invalid-url-pattern-context-crash-expected.txt: Added.
* LayoutTests/fast/url/invalid-url-pattern-context-crash.html: Added.
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::match const):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
(WebCore::URLPatternUtilities::URLPatternComponent::createComponentMatchResult const):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.h:

Canonical link: <a href="https://commits.webkit.org/294350@main">https://commits.webkit.org/294350@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ad2571e2de1a6be625cbba4848c2b9705147bd8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101353 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11323 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106507 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51983 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21324 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29510 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77198 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34233 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104360 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16453 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91548 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57539 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16273 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9558 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51331 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86154 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9634 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108857 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28481 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20956 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86173 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87750 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85729 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30455 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8167 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/22596 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16522 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28411 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33693 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28223 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31543 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -111,6 +111,8 @@ URLPatternComponentResult URLPatternComponent::createComponentMatchResult(Script
 
     auto globalObject = context.globalObject();
+    if (!globalObject)
+        return URLPatternComponentResult { emptyString(), WTFMove(groups) };
```
I wonder what other browsers do, maybe they throw an exception.
It would be good to check in the test the various values of the created `URLPattern` and validate the results are matching Chrome.

Also, do we need that check in all `createComponentMatchResult`, or is it sufficient to have one single `JSGlobalObject` null check in `URLPattern::match`?

pratiksha-99 replied:
I just updated to check in `URLPattern::match` for empty context. Tests looked ok. Let me know what you think.

pratiksha-99 replied:
Hmm..looks like more tests failed, might have to adjust existing tests. I will check what Chrome does for this case.

pratiksha-99 replied:
Chrome's implementation of the URLPattern API does not appear to throw an exception when the execution context is invalid. So returning just null should be ok, lets wait for EWS.

rniwa commented on LayoutTests/fast/url/invalid-url-pattern-context-crash.html:
```diff
@@ -0,0 +1,18 @@
+
+function runTest() {
+    try {
+      frameElement.remove();
```
Please indent this with 4 spaces like the rest of the file for consistency.

youennf commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -440,6 +440,9 @@ ExceptionOr<std::optional<URLPatternResult>> URLPattern::match(ScriptExecutionCo
+
+    if (!context.globalObject())
+        return { std::nullopt };
     result.protocol = m_protocolComponent.createComponentMatchResult(context, WTFMove(protocol), protocolExecResult);
```
I would tend to pass the globalObject as a reference here, as createComponentMatchResult only uses a globalObject. Something like:
```
auto* globalObject = context.globalObject();
if (!globalObject)
    return  { };
result.protocol = m_protocolComponent.createComponentMatchResult(globalObject, WTFMove(protocol), protocolExecResult);
```

pratiksha-99 replied:
Done.

youennf commented on LayoutTests/fast/url/invalid-url-pattern-context-crash.html:
```diff
@@ -0,0 +1,18 @@
+        try {
+          frameElement.remove();
+      } catch (e) {}
+      document.getElementById('htmlObjectElement').data = document.documentURI;
```
I do not understand the test. `frameElement` is probably null here.
I would expect the following:
1. load an iframe, named `myframe`.
2. When the iframe is loaded, run:
```
const contentWindow = myframe.contentWindow;
myframe.remove();
(new contentWindow.URLPattern()).test();
```

pratiksha-99 replied:
Yeah, but the above logic doesnt repro the crash unfortunately. Layout system maybe skipping some steps because it thinks the document is detached, but somehow subsequent JS calls access global state/layout, causing a crash.

youennf reviewed (commented):
Code change LGTM, a nit could be done though it could be done as a follow-up if we want to minimise the size of the patch.
I did not r+ it as I am not sure the test covers the change.

pratiksha-99 committed:
Nullptr crash in URLPatternUtilities::URLPatternComponent::createComponentMatchResult
https://bugs.webkit.org/show_bug.cgi?id=291924
rdar://149686034

Reviewed by Youenn Fablet.

Added a check for null global object context in case it is broken or detached.

* LayoutTests/fast/url/invalid-url-pattern-context-crash-expected.txt: Added.
* LayoutTests/fast/url/invalid-url-pattern-context-crash.html: Added.
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::URLPattern::match const):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
(WebCore::URLPatternUtilities::URLPatternComponent::createComponentMatchResult const):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.h:

Canonical link: https://commits.webkit.org/294350@main
